alibi/explainers/similarity/base.py

```diff
@@ -0,0 +1,174 @@
+                                   Callable[[torch.Tensor, torch.Tensor], torch.Tensor]]''',
+                 sim_fn: Callable[[np.ndarray, np.ndarray], np.ndarray],
+                 precompute_grads: bool = False,
+                 backend: Framework = Framework.TENSORFLOW,
+                 device: 'Union[int, str, torch.device, None]' = None,
+                 meta: Optional[dict] = None,
+                 ):
+        """Constructor
+
+        Parameters
+        ----------
+        predictor
+            Model to be explained.
+        loss_fn
+            Loss function.
+        sim_fn
+            Similarity function. Takes two inputs and returns a similarity value.
+        precompute_grads
+            Whether to precompute and store the gradients when fitting.
+        backend
+            Deep learning backend.
+        device
+            Device to be used. Will default to the same device the backend defaults to.
+        meta
+            Metadata specific to explainers that inherit from this class. Should be initialized in the child class and
+            passed in here. Is used in the `__init__` of the base Explainer class.
+        """
+
+        # Select backend.
+        self.backend = _select_backend(backend)
+        self.backend.set_device(device)  # type: ignore
+
+        self.predictor = predictor
+        self.loss_fn = loss_fn
+        self.sim_fn = sim_fn
+        self.precompute_grads = precompute_grads
+
+        meta = {} if meta is None else meta
+        super().__init__(meta=meta)
+
+    def fit(self,
+            X_train: np.ndarray,
+            Y_train: np.ndarray) -> "Explainer":
+        """Fit the explainer. If ``self.precompute_grads == True`` then the gradients are precomputed and stored.
+
+        Parameters
+        ----------
+        X_train
+            Training data.
+        Y_train
+            Training labels.
+
+        Returns
+        -------
+        self
+            Returns self.
+        """
+        self.X_train: np.ndarray = X_train
+        self.Y_train: np.ndarray = Y_train
+        self.X_dims: Tuple = self.X_train.shape[1:]
+        self.Y_dims: Tuple = self.Y_train.shape[1:]
+        self.grad_X_train: np.ndarray = np.array([])
+
+        # compute and store gradients
+        if self.precompute_grads:
+            grads = []
+            for X, Y in tqdm(zip(self.X_train, self.Y_train)):
+                grad_X_train = self._compute_grad(X[None], Y[None])
+                grads.append(grad_X_train[None])
+            self.grad_X_train = np.concatenate(grads, axis=0)
+        return self
+
+    def _verify_fit(self) -> None:
+        """Verify that the explainer has been fitted.
+
+        Raises
+        ------
+        ValueError
+            If the explainer has not been fitted.
+        """
+
+        if not hasattr(self, 'X_train') or not hasattr(self, 'Y_train'):
+            raise ValueError('Training data not set. Call `fit` and pass training data first.')
```

Comment:
shall we have a special case when `precompute_grads` is `True`. In that case, we need to check if the gradients are also precomputed.

Reply:
I don't think this can happen because the gradients are computed in the same step in which you set the training and label data? I guess a user might initially set up the explainer to not precompute gradients, then set the training data, and then set precompute gradients to true. Is this a case we have to deal with though? @jklaise, what do you think?

Reply:
Not sure how that can happen unless they fiddle with the attributes directly (which they shouldn't)?